### PR TITLE
Track individual bonus parts; fixes #65

### LINKED ISSUE #65: Record scores on individual bonus parts
Especially now that NAQT's scoresheet does so, we should represent scoring on individual bonus parts rather than for the bonus as a whole.

Here's my first stab at how to represent it: In `MatchQuestion`, remove `bonus_points` and `bounceback_bonus_points`, and change to `bonus_parts`, which is an array of `BonusPart` objects, and may be omitted if the question type doesn't award a bonus.

Each `BonusPart` has `points` (number; required) and `bounceback_points` (number; may be omitted if bonuses don't bounce back)

For "normal bonuses" in which there are distinct parts in a definite order, each part gets its own `BonusPart`, and — at least in every format I'm familiar with — `bounceback_points` can only be nonzero if points is zero. For weird cases like "name any _n_ of the _m_ things", just treat the whole bonus as a single `BonusPart`, and then you might report something like `{points: 20, bounceback_points: 10}`.

### PR DESCRIPTION
**Pull request by jonahgreenthal**

### CODE REVIEW

File: match.md [hunk 1]
```diff
@@ -241,3 +248,35 @@ If a `Match` includes `Lineup` objects on its `MatchTeam` objects and includes `
...
+<table class="fields"><tbody>
+  <tr class="required">
+    <th>controlled_points</th>
+    <th class="type">Number</td>
+    <td>Number of points earned by the controlling team. In all formats with which we are familiar, this must be a nonnegative integer.</td>
+  </tr>
+  <tr class="optional">
+    <th>bounceback_points</th>
+    <th class="type">Number</td>
+    <td>Number of points earned by the non-controlling team. May be omitted if bonuses do not bounce back, but should not be omitted merely because the non-controlling team earned no points. In all formats with which we are familiar, this must be a nonnegative integer, and can only be positive if `controlled_points` is `0`.</td>
```
Comment (puls) [inline match.md]: I would strengthen this statement and remove "In all formats with which we are familiar".

File: match.md [hunk 2]
```diff
@@ -213,15 +213,22 @@ If a `Match` includes `Lineup` objects on its `MatchTeam` objects and includes `
...
   <tr class="optional">
     <th>bounceback_bonus_points</th>
     <td class="type">Number</td>
-    <td>How many bonus points were awarded when this question's bonus bounced back. May be omitted if the question type doesn't award a bonus or if bonuses don't bounce back.</td>
+    <td>How many bonus points were awarded when this question's bonus bounced back. May be omitted if `bonus` is present, if the question type doesn't award a bonus, or if bonuses don't bounce back.</td>
+  </tr>
+  <tr class="optional">
+    <th>bonus</th>
+    <td class="type">`MatchQuestionBonus`</td>
+    <td>Information about the bonus that was read for this question. May be omitted if `bonus_points` is present, or if no bonus was read for this question, but should not be omitted merely because no points were earned on the bonus.</td>
```
Comment (puls) [inline match.md]: Maybe make `bonus` and `bonus_points` be mutually exclusive?

Comment (jonahgreenthal) [inline match.md]: Isn't that what line 230 says?

**Comment by jonahgreenthal** [comment]
What needs to happen before we can merge this?

**Comment by puls** [comment]
I left a comment above; usually I don't merge PRs until everything is addressed, either by changing the code or by replying to say you'd rather not change it.

File: match.md (hunk 1 above)
Comment (jonahgreenthal) [inline match.md]: I tweaked this in 3d2d9c300288e90fefc8a880db9013543cfc538f. I do prefer to hedge with "in all formats with which we are familiar"; for instance, I can imagine some IESA-esque format adding a penalty for illegal conferring on a bonus.

**Comment by jonahgreenthal** [comment]
What hasn't been addressed?

**Pull request merged**